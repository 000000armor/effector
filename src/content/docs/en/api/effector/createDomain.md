---
title: createDomain
description: createDomain is a method for creating a domain
---

Creates a [domain](/api/effector/Domain.md)

```typescript
createDomain(name?)
```

**Arguments**

1. `name`? (_string_): domain name

**Returns**

[_Domain_](/api/effector/Domain.md): New domain

### Example

```js
import { createDomain } from "effector";

const domain = createDomain(); // Unnamed domain
const httpDomain = createDomain("http"); // Named domain

const statusCodeChanged = httpDomain.createEvent();
const downloadFx = httpDomain.createEffect();
const apiDomain = httpDomain.createDomain(); // nested domain
const $data = httpDomain.createStore({ status: -1 });
```

[Try it](https://share.effector.dev/GMpjINHa)
