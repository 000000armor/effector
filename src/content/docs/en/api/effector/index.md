---
id: effector
title: API Reference
---

# List of the methods

## Unit creators

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Unit definitions

- [Event](./Event)
- [Effect](./Effect)
- [Store](./Store)
- [Domain](./Domain)
- [Scope](./Scope)

## Common methods

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Plugins

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Utilities

- [is](./is)

## Low level api

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
