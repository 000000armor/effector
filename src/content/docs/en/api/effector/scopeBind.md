---
title: scopeBind
description: scopeBind is a method to bind unit to scope
---

## Formulae {#scopeBind-formulae}

```ts
scopeBind<T>(event: Event<T>): (payload: T) => void
```

A method to bind event or effect to a [Scope](/en/api/effector/Scope) to be called later. effector supports imperative event calling inside watchers, but sometimes you should bind events to the scope explicitly, e.g. when calling events within a setTimeout/setInterval callbacks.

### Arguments {#scopeBind-arguments}

1. `event` [_Event_](/en/api/effector/Event) to be bound to the scope

## Example {#scopeBind-example}

We are going to call `changeLocation` inside `history.listen` callback so there is no way for effector to associate event with corresponding scope, and we should explicitly bind event to scope using `scopeBind`.

```js
const installHistory = app.createEvent<any>()
const changeLocation = app.createEvent<string>()

installHistory.watch(history => {

  const locationUpdate = scopeBind(changeLocation)
  history.listen(location => {
    locationUpdate(location.pathname)
  })
})
```

[See full example](https://github.com/effector/effector/blob/master/examples/react-ssr/src/app.tsx#L128)
