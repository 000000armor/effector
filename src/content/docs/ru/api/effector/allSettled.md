---
title: allSettled
lang: ru
---

::: info Нет перевода
Эта статья еще не переведена на русский язык. <br/>
Пожалуйста, используйте ссылку внизу "Edit this page", чтобы добавить перевод.<br/>
[Прочитайте оригинальную статью](/api/effector/allSettled.md) на английском.
:::
