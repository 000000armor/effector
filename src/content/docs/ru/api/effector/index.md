---
title: Справочник по API
description: Перечень основных методов API, по группам
lang: ru
---

Перечень основных методов API, по группам

## Создание юнитов

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Типы юнитов

- [_Event_](./Event)
- [_Effect_](./Effect)
- [_Store_](./Store)
- [_Domain_](./Domain)
- [_Scope_](./Scope)

## Основные методы библиотеки

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Плагины для компилятора

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Служебные функции

- [is](./is)

## Низкоуровневый API

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
