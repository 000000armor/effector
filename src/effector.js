//@flow

import type {Store} from 'redux'
import {Stream, from} from 'most'
import {async as subject, type Subject} from 'most-subject'

import type {ID, Tag, Domain, Effect, Event} from './index.h'
import {counter, toTag} from './index.h'

export function createDomain<State>(
  store: Store<State>,
  domainName: string = '~'
): Domain<State> {
  const redux: any = store
  const state$: Stream<State> = from(redux)
    .map(() => store.getState())
    .multicast()
  return DomainConstructor(
    domainName,
    redux.dispatch,
    store.getState,
    state$,
  )
}

export function domain<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
): Domain<State> {
  return DomainConstructor(
    domainName,
    dispatch,
    getState,
    state$,
  )
}


const nextPayloadID = counter()
const nextEventID = counter()

function DomainConstructor<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events = new Map
): Domain<State> {
  return {
    effect<Params, Done, Fail>(
      name: string
    ): Effect<Params, Done, Fail, State> {
      return EffectConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    },
    domain(name: string) {
      return DomainConstructor(
        [domainName, name].join('/'),
        dispatch,
        getState,
        state$,
        events,
      )
    },
    event<Payload>(
      name: string
    ): Event<Payload, State> {
      return EventConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    }
  }
}

function EffectConstructor<State, Params, Done, Fail>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Effect<Params, Done, Fail, State> {
  const handlers = new Set
  const eventID = nextEventID()
  const nextSeq = counter()
  const action$: Subject<Params> = subject()
  const getType = () => toTag(domainName, name)
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  let using: (params: Params) => Promise<Done> = never
  function use(thunk: (params: Params) => Promise<Done>) {
    using = thunk
  }
  const create = (payload: Params) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const effect = (params: Params) => {
    let run = () => {
      run = () => {}
      using(params).then(
        result => {
          const data = {params, result}
          dispatch(done(data))
          resolve(data)
        },
        error => {
          const data = {params, error}
          dispatch(fail(data))
          reject(data)
        },
      )
    }
    const runner = (dispatchHook) => {
      if (canDispatch) {
        const dispatcher = dispatchHook
          ? dispatchHook
          : dispatch
        canDispatch = false
        dispatcher(result)
        for (const handler of handlers) {
          handler(params, getState())
        }
        action$.next(params)
        run()
      }
    }
    let resolve = (val) => {}
    let reject = (val) => {}
    const doneP: Promise<{params: Params, result: Done}> = new Promise(rs => resolve = rs)
    const failP: Promise<{params: Params, error: Fail}> = new Promise(rs => reject = rs)
    const toSend = Promise.resolve(params)
    const result = create(params)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        runner(dispatchHook)
        return toSend
      },
      done() {
        runner()
        return doneP
      },
      fail() {
        runner()
        return failP
      },
      promise() {
        runner()
        return Promise.race([
          doneP,
          failP.then(data => Promise.reject(data)),
        ])
      }
    }
  }
  const done: Event<{params: Params, result: Done}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'done'].join(' ')
  )
  const fail: Event<{params: Params, error: Fail}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'fail'].join(' ')
  )
  function watch<R>(fn: (params: Params, state: State) => R) {
    handlers.add((params, state) => {
      const result = fn(params, state)
      safeDispatch(result, dispatch)
    })
  }
  function epic<R>(
    handler: (
      data$: Stream<Params>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  effect.use = use
  effect.getType = getType
  effect.done = done
  effect.fail = fail
  effect.watch = watch
  effect.epic = epic
  return effect
}

function safeDispatch(data, dispatch) {
  if (
    typeof data === 'object'
    && data != null
    && typeof data.type === 'string'
  ) {
    dispatch(data)
  }
}

function EventConstructor<State, Payload>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Event<Payload, State> {
  const eventID = nextEventID()
  const nextSeq = counter()
  const getType = () => toTag(domainName, name)
  const handlers = new Set
  function watch<R>(
    fn: (params: Payload, state: State) => R
  ) {
    handlers.add(fn)
  }
  const action$: Subject<Payload> = subject()
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  function epic<R>(
    handler: (
      data$: Stream<Payload>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  const create = (payload: Payload) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const eventInstance = (payload: Payload) => {
    const toSend = Promise.resolve(payload)
    const result = create(payload)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        if (canDispatch) {
          const dispatcher = dispatchHook
            ? dispatchHook
            : dispatch
          canDispatch = false
          dispatcher(result)
          for (const handler of handlers) {
            handler(payload, getState())
          }
          action$.next(payload)
        }
        return toSend
      },
    }
  }
  eventInstance.getType = getType
  //$off
  eventInstance.toString = getType
  eventInstance.watch = watch
  eventInstance.epic = epic

  events.set(getType(), eventInstance)
  return eventInstance
}


function never(props: any): Promise<any> {
  console.warn(`


  Running an effect without implementation

`, props)
  return new Promise(() => {})
}


