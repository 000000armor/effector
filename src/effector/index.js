//@flow

/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball of wibbly wobbly time-y wimey stuff
 */

export {combine, combine as createStoreObject} from './combine'
export {sample} from './sample'

export type {Domain} from './domain'
export {createDomain} from './domain'

export type {Event} from './event'
export {createEvent, merge, forward, fromObservable} from './event'

export type {Effect} from './effect'
export {createEffect} from './effect'

export type {Store} from './store'
export {
  createStore,
  setStoreName,
  extract,
  createApi,
  restore,
  restoreEvent,
  restoreEffect,
  restoreObject,
  withProps,
} from './store'

export {
  Kind,
  clearNode,
  createGraph as createNode,
  step,
  traverse,
} from './stdlib'
export {launch} from './kernel'
export type {kind} from './stdlib'

export {
  invariant,
  warning,
  isUnit,
  isStore,
  isEvent,
  isEffect,
  isDomain,
  is,
} from './validate'
export {version} from 'effector/flags'

//eslint-disable-next-line
export * as blocks from './blocks'
