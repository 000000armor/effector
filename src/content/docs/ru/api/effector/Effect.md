---
title: Effect
keywords:
  - effect
  - side-effect
  - unit
  - эффект
  - сайд-эффект
  - юнит
description: Effect, его методы и свойства
lang: ru
---

_Effect (эффект)_ это контейнер для сайд-эффектов, возможно асинхронных. В комплекте имеет ряд заранее созданных эвентов и сторов, облегчающих стандартные действия. Является [юнитом](/ru/explanation/glossary#common-unit)

Эффекты можно вызывать как обычные функции (_императивный вызов_) а также подключать их и их свойства в различные методы api включая [sample](/ru/api/effector/sample), [guard](/ru/api/effector/guard) и [split](/ru/api/effector/split) (_декларативное подключение_). При императивном вызове принимают максимум один аргумент и всегда возвращают промис, отражающий ход выполнения сайд-эффекта


# Методы {#effect-methods}

## `use(handler)` {#effect-use-handler}

Определяет имплементацию эффекта: функцию, которая будет вызвана при срабатывании. Используется для случаев когда имплементация не установлена [при создании](/ru/api/effector/createEffect) или когда требуется изменение поведения эффекта при тестировании

Если на момент вызова эффект уже имел имплементацию, то она будет заменена на новую

:::tip{title="статья от автора"}
[Testing api calls with effects and stores](https://www.patreon.com/posts/testing-api-with-32415095)
:::

:::info
Нужно предоставить имплементацию либо через use, либо через [createEffect](/ru/api/effector/createEffect), иначе при вызове эффекта возникнет ошибка "no handler used in _%effect name%_"
:::

### Формула {#effect-use-handler-formulae}

```ts
declare const fx: Effect<T, S>;

fx.use(/*handler*/ (params: T) => S | Promise<S>);
```

### Аргументы {#effect-use-handler-arguments}

1. **`handler`**: `(params: T) => S | Promise<S>`

   Функция-имплементация эффекта. Может быть как синхронной, так и асинхронной

   **Аргументы**

    1. **`params`**: Данные, с которыми был вызван эффект

   **Возвращает**

   Результат выполнения эффекта в виде значения, либо в виде промиса со значением

### Возвращает {#effect-use-handler-return}

Текущий эффект

:::info
Если значение имплементации известно сразу, то оптимальнее использовать `createEffect(handler)`

`createEffect().use(handler)` это антипаттерн, который ухудшает вывод типов
:::

### Пример {#effect-use-handler-example}

```js
import { createEffect } from "effector";

const fetchUserReposFx = createEffect();

// ....

fetchUserReposFx.use(async ({ name }) => {
  console.log("fetchUserReposFx вызван для github пользователя", name);

  const url = `https://api.github.com/users/${name}/repos`;
  const req = await fetch(url);
  return req.json();
});

await fetchUserReposFx({ name: "zerobias" });
// => fetchUserReposFx вызван для github пользователя zerobias
```

[Запустить пример](https://share.effector.dev/Vp8tPzBh)

## `use.getCurrent()` {#effect-use-getCurrent}

Метод для получения текущей имплементации эффекта. Используется для тестирования

Если у эффекта ещё не была установлена имплементация, то будет возвращена функция по умолчанию, при срабатывании она [выбрасывает ошибку](https://share.effector.dev/8PBjt3TL)

### Формула {#effect-getCurrent-formulae}

```ts
declare const fx: Effect<P, D>

fx.use.getCurrent()
-> (params: P) => D
```

### Возвращает {#effect-getCurrent-return}

Функцию-имплементацию эффекта, которая была установлена через [createEffect](/ru/api/effector/createEffect) или с помощью метода [use](#use)

### Пример {#effect-getCurrent-example}

```js
const handlerA = () => "A";
const handlerB = () => "B";

const fx = createEffect(handlerA);

console.log(fx.use.getCurrent() === handlerA);
// => true

fx.use(handlerB);
console.log(fx.use.getCurrent() === handlerB);
// => true
```

[Запустить пример](https://share.effector.dev/CM6hgtOM)

## `watch(watcher)` {#effect-watch-watcher}

Вызывает дополнительную функцию с сайд-эффектами при каждом срабатывании эффекта

:::info
По мере усложнения логики проекта оптимальнее заменить на комбинацию дополнительного эффекта и [sample](/ru/api/effector/sample)
:::

### Формула {#effect-watch-watcher-formulae}

```ts
declare const fx: Effect<T, any>

fx.watch(/*watcher*/ (data: T) => any)
-> Subscription
```

### Аргументы {#effect-watch-watcher-arguments}

1. **`watcher`**: `(data: T) => any`

   Функция с сайд-эффектами, в качестве первого аргумента получает значение с которым был вызван эффект. Возвращаемое значение не используется

### Возвращает {#effect-watch-watcher-return}

[_Subscription_](/ru/explanation/glossary#subscription): Функция отмены подписки, после её вызова `watcher` перестаёт получать обновления и удаляется из памяти. Повторные вызовы функции отмены подписки не делают ничего

### Пример {#effect-watch-watcher-example}

```js
import { createEffect } from "effector";

const fx = createEffect((params) => params);

fx.watch((params) => {
  console.log("эффект вызван с аргументом", params);
});

await fx(10);
// => эффект вызван с аргументом 10
```

[Запустить пример](https://share.effector.dev/iNb7YIdV)

## `prepend(fn)` {#effect-prepend-fn}

Создаёт событие-триггер для преобразования данных _перед_ запуском эффекта. По сравнению с [map](#map), работает в обратном направлении

### Формула {#effect-prepend-fn-formulae}

```ts
declare const fx: Effect<S, any>

const trigger = fx.prepend(/*fn*/(data: T) => S)
-> Event<T>
```

При вызове события `trigger`, функция-обработчик `fn` будет вызвана с поступившими данными, после чего `fx` будет вызван с результатом вычислений

```

    trigger -> fn -> fx

```

### Аргументы {#effect-prepend-fn-arguments}

1.  **`fn`**: `(data: T) => S`

    Функция-обработчик, которая будет вычислять данные для передачи в `fx` на основе данных события `trigger`. [Должна быть **чистой**](/ru/explanation/glossary#purity)

    **Аргументы**

    - **`data`**: Данные с которыми сработало событие `trigger`

    **Возвращает**

    Данные для передачи в `fx`

### Возвращает {#effect-prepend-fn-return}

Новое событие


## `map(fn)` {#effect-map-fn}

Создает производное событие на основе данных эффекта

### Формула {#effect-map-fn-formulae}

```ts
declare const fxA: Effect<T, any>

const eventB = fxA.map(/*fn*/(data: T) => S)
-> Event<S>
```

При вызове `fxA`, функция-обработчик `fn` будет вызвана с поступившими данными, после чего производное событие `eventB` будет вызвано с результатом вычислений

```

    fxA -> fn -> eventB

```

### Аргументы {#effect-map-fn-arguments}

1.  **`fn`**: `(data: T) => S`

    Функция-обработчик, которая будет вычислять данные для передачи в производное событие `eventB` на основе данных из `fxA`. [Должна быть **чистой**](/ru/explanation/glossary#purity)

    **Аргументы**

    - **`data`**: Данные с которыми сработал эффект `fxA`

    **Возвращает**

    Данные для передачи в производное событие `eventB`

### Возвращает {#effect-map-fn-return}

Новое, производное событие

### Пример {#effect-map-fn-example}

```js
import { createEffect } from "effector";

const updateUserFx = createEffect(({ name, role }) => {});
const userNameUpdate = updateUserFx.map(({ name }) => name);
const userRoleUpdate = updateUserFx.map(({ role }) => role.toUpperCase());

userNameUpdate.watch((name) => {
  console.log(`Началось изменение имени пользователя на ${name}`);
});
userRoleUpdate.watch((role) => {
  console.log(`Началось изменение роли пользователя на ${role}`);
});

await updateUserFx({ name: "john", role: "admin" });
// => Началось изменение имени пользователя на john
// => Началось изменение роли пользователя на ADMIN
```

[Запустить пример](https://share.effector.dev/4UFLTo5p)

# Свойства {#effect-properties}

## `done` {#effect-done}

[Событие](/ru/api/effector/Event), которое срабатывает с результатом выполнения эффекта и аргументом, переданным при вызове

### Формула {#effect-done-formulae}

```ts
declare const fx: Effect<P, D>

fx.done
-> Event<{params: P; result: D}>
```

:::info{title="Вызов вручную запрещён"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-done-example}

```js
import { createEffect } from "effector";

const fx = createEffect((value) => value + 1);

fx.done.watch(({ params, result }) => {
  console.log("Вызов с аргументом", params, "завершён со значением", result);
});

await fx(2);
// => Вызов с аргументом 2 завершён со значением 3
```

[Запустить пример](https://share.effector.dev/tnSg24Ca)


## `doneData` {#effect-doneData}

[Событие](/ru/api/effector/Event), которое срабатывает с результатом выполнения эффекта

### Формула {#effect-doneData-formulae}

```ts
declare const fx: Effect<any, D>

fx.doneData
-> Event<D>
```

:::info{title="Вызов вручную запрещён"}
Это свойство управляется самим эффектом
:::

:::info
Добавлено в effector 20.12.0
:::

### Пример {#effect-doneData-example}

```js
import { createEffect } from "effector";

const fx = createEffect((value) => value + 1);

fx.doneData.watch((result) => {
  console.log(`Эффект успешно выполнился, вернув ${result}`);
});

await fx(2);
// => Эффект успешно выполнился, вернув 3
```

[Запустить пример](https://share.effector.dev/KexWC7GO)

## `fail` {#effect-fail}

[Событие](/ru/api/effector/Event), которое срабатывает с ошибкой, возникшей при выполнении эффекта и аргументом, переданным при вызове

### Формула {#effect-fail-formulae}

```ts
declare const fx: Effect<P, any, E>

fx.fail
-> Event<{params: P; error: E}>
```

:::info{title="Вызов вручную запрещён"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-fail-example}

```js
import { createEffect } from "effector";

const fx = createEffect(async (value) => {
  throw Error(value - 1);
});

fx.fail.watch(({ params, error }) => {
  console.log("Вызов с аргументом", params, "завершился с ошибкой", error.message);
});

fx(2);
// => Вызов с аргументом 2 завершился с ошибкой 1
```

[Запустить пример](https://share.effector.dev/5xHVmzIJ)

## `failData` {#effect-failData}

:::info
Добавлено в effector 20.12.0
:::

[Событие](/ru/api/effector/Event), которое срабатывает с ошибкой, возникшей при выполнении эффекта

### Формула {#effect-failData-formulae}

```ts
declare const fx: Effect<any, any, E>

fx.failData
-> Event<E>
```

:::info{title="Вызов вручную запрещён"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-failData-example}

```js
import { createEffect } from "effector";

const fx = createEffect(async (value) => {
  throw Error(value - 1);
});

fx.failData.watch((error) => {
  console.log(`Вызов завершился с ошибкой ${error.message}`);
});

fx(2);
// => Вызов завершился с ошибкой 1
```

[Запустить пример](https://share.effector.dev/i5ktYSqM)

## `finally` {#effect-finally}

:::info
Добавлено в effector 20.0.0
:::

[Событие](/ru/api/effector/Event), которое срабатывает при завершении эффекта с подробной информацией об аргументах, результатах и статусе выполнения

### Формула {#effect-finally-formulae}

```ts
declare const fx: Effect<P, D, E>

fx.finally
-> Event<
  | {status: 'done'; params: P; result: D}
  | {status: 'fail'; params: P; error: E}
>
```

:::info{title="Вызов вручную запрещён"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-finally-example}

```js
import { createEffect } from "effector";

const fetchApiFx = createEffect(async ({ time, ok }) => {
  await new Promise((resolve) => setTimeout(resolve, time));
  if (ok) return `${time} ms`;
  throw Error(`${time} ms`);
});

fetchApiFx.finally.watch((value) => {
  switch (value.status) {
    case "done":
      console.log("Вызов с аргументом", value.params, "завершён со значением", value.result);
      break;
    case "fail":
      console.log("Вызов с аргументом", value.params, "завершён с ошибкой", value.error.message);
      break;
  }
});

await fetchApiFx({ time: 100, ok: true });
// => Вызов с аргументом {time: 100, ok: true}
//    завершён со значением 100 ms

fetchApiFx({ time: 100, ok: false });
// => Вызов с аргументом {time: 100, ok: false}
//    завершён с ошибкой 100 ms
```

[Запустить пример](https://share.effector.dev/Oqvy2x35)


## `pending` {#effect-pending}

[Стор](/ru/api/effector/Store), который показывает, что эффект находится в процессе выполнения

### Формула {#effect-pending-formulae}

```ts
declare const fx: Effect<any, any>

fx.pending
-> Store<boolean>
```

Это свойство избавляет от необходимости писать подобный код:

```js
import { createEffect, createStore } from "effector";

const requestFx = createEffect();

const $isRequestPending = createStore(false)
  .on(requestFx, () => true)
  .on(requestFx.done, () => false)
  .on(requestFx.fail, () => false);
```

:::info{title="Изменение значения вручную запрещено"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-pending-example}

#### Отображение индикатора загрузки с React {#effect-pending-example-react}

```jsx
import React from "react";
import ReactDOM from "react-dom";
import { createEffect } from "effector";
import { useStore } from "effector-react";

const fetchApiFx = createEffect(async (ms) => {
  await new Promise((resolve) => setTimeout(resolve, ms));
});

fetchApiFx.pending.watch(console.log);
// => false

const App = () => {
  const loading = useStore(fetchApiFx.pending);
  return <div>{loading ? "Загрузка..." : "Загрузка завершена"}</div>;
};

ReactDOM.render(<App />, document.getElementById("root"));

fetchApiFx(1000);
// => true
// => false
```

[Запустить пример](https://share.effector.dev/e9y5uETf)

## `inFlight` {#effect-inFlight}

:::info
Добавлено в effector 20.11.0
:::

[Стор](/ru/api/effector/Store), который показывает число запущенных эффектов, которые находятся в процессе выполнения. Используется для ограничения числа одновременных запросов

### Формула {#effect-inFlight-formulae}

```ts
declare const fx: Effect<any, any>

fx.inFlight
-> Store<number>
```

Это свойство избавляет от необходимости писать подобный код:

```js
import { createEffect, createStore } from "effector";

const requestFx = createEffect();

const $requestsInFlight = createStore(0)
  .on(requestFx, (n) => n + 1)
  .on(requestFx.done, (n) => n - 1)
  .on(requestFx.fail, (n) => n - 1);
```

:::info{title="Изменение значения вручную запрещено"}
Это свойство управляется самим эффектом
:::

### Пример {#effect-inFlight-example}

```js
import { createEffect } from "effector";

const fx = createEffect(async () => {
  await new Promise((resolve) => setTimeout(resolve, 500));
});

fx.inFlight.watch((amount) => {
  console.log("выполняется запросов:", amount);
});
// => выполняется запросов: 0

const req1 = fx();
// => выполняется запросов: 1

const req2 = fx();
// => выполняется запросов: 2

await Promise.all([req1, req2]);

// => выполняется запросов: 1
// => выполняется запросов: 0
```

[Запустить пример](https://share.effector.dev/ADD0M4NV)

## `shortName` {#effect-shortName}

Имя эффекта. Задаётся либо явно, через поле `name` в [createEffect](/ru/api/effector/createEffect), либо автоматически через [babel plugin](/ru/api/effector/babel-plugin). Используется для обработки сущностей программно, например при использовании [хуков домена](/ru/api/effector/Domain#onCreateEffect)

### Формула {#effect-shortName-formulae}

```ts
declare const fx: Effect<any, any>

fx.shortName
-> string
```

## `sid` {#effect-sid}

Стабильный идентификатор эффекта. Задаётся автоматически через [Babel plugin](/ru/api/effector/babel-plugin)

### Формула {#effect-sid-formulae}

```ts
declare const fx: Effect<any, any>

fx.sid
-> string | null
```