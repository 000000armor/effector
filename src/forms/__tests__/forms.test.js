//@flow

import faker from 'faker'
import {type Effect, createEffect, createEvent, restoreEvent} from 'effector'
import {createFormApi} from '../'

describe('createFormApi', () => {
  test('submitEffect', () => {
    const submitEffect: Effect<
      {|
        firstName: string,
        lastName: string,
      |},
      string,
      number,
    > = createEffect()
    submitEffect.use(async() => 'foo')

    submitEffect.watch(params => expect(params).toMatchSnapshot())

    const form = createFormApi({
      fields: {
        firstName: '',
        lastName: '',
      },
      submitEffect,
    })
    form.submitted.watch(a => expect(a).toMatchSnapshot('submitted'))
    form.api.firstName('bar')
    form.handleSubmit()
  })

  test('submitEvent', () => {
    const submitEvent = createEvent<{|
      firstName: string,
      lastName: string,
    |}>()
    submitEvent.watch(params => expect(params).toMatchSnapshot())
    const form = createFormApi({
      fields: {
        firstName: '',
        lastName: '',
      },
      submitEvent,
    })
    form.submitted.watch(a => expect(a).toMatchSnapshot('submitted'))
    form.api.firstName('foo')
    form.handleSubmit()
  })

  test('validate', () => {
    const submitEvent = createEvent<{|
      firstName: string,
      lastName: string,
    |}>()
    submitEvent.watch(params => expect(params).toMatchSnapshot())
    const form = createFormApi({
      fields: {
        firstName: '',
        lastName: '',
      },
      validate({firstName, lastName}, data) {
        if (firstName.length < 6) return {firstName: 'First name is too short.'}
        return {}
      },
      submitEvent,
    })
    form.submitted.watch(a => expect(a).toMatchSnapshot('submitted'))
    form.handleSubmit()
    form.api.firstName('foobar')
    form.handleSubmit()
  })

  test('initial values', () => {
    faker.seed(0xdeadbeef)
    const submitEvent = createEvent<{|
      firstName: string,
      lastName: string,
    |}>()
    const generate = createEvent()
    const initialValues = restoreEvent(generate, {
      firstName: faker.name.firstName(),
      lastName: faker.name.lastName(),
    }).map(() => ({
      firstName: faker.name.firstName(),
      lastName: faker.name.lastName(),
    }))
    const form = createFormApi({
      initialValues,
      fields: {
        firstName: '',
        lastName: '',
      },
      submitEvent,
    })
    initialValues.watch(values =>
      expect(values).toMatchSnapshot('initial values'),
    )
    form.values.watch(values => expect(values).toMatchSnapshot('values'))
    generate(1)
    form.api.firstName('Alan')
    generate(2)
    generate(3)
  })
})
