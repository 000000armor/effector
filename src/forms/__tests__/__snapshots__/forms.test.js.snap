// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFormApi initial values: initial values 1`] = `
Object {
  "firstName": "Ollie",
  "lastName": "Keebler",
}
`;

exports[`createFormApi initial values: initial values 2`] = `
Object {
  "firstName": "Lola",
  "lastName": "Conn",
}
`;

exports[`createFormApi initial values: initial values 3`] = `
Object {
  "firstName": "Dagmar",
  "lastName": "Howell",
}
`;

exports[`createFormApi initial values: initial values 4`] = `
Object {
  "firstName": "Nelson",
  "lastName": "Jones",
}
`;

exports[`createFormApi initial values: values 1`] = `
Object {
  "firstName": "Ollie",
  "lastName": "Keebler",
}
`;

exports[`createFormApi initial values: values 2`] = `
Object {
  "firstName": "Lola",
  "lastName": "Conn",
}
`;

exports[`createFormApi initial values: values 3`] = `
Object {
  "firstName": "Alan",
  "lastName": "Conn",
}
`;

exports[`createFormApi submitEffect 1`] = `
Object {
  "firstName": "bar",
  "lastName": "",
}
`;

exports[`createFormApi submitEffect: submitted 1`] = `false`;

exports[`createFormApi submitEffect: submitted 2`] = `true`;

exports[`createFormApi submitEffect: submitted 3`] = `false`;

exports[`createFormApi submitEvent 1`] = `
Object {
  "firstName": "foo",
  "lastName": "",
}
`;

exports[`createFormApi submitEvent: submitted 1`] = `false`;

exports[`createFormApi submitEvent: submitted 2`] = `true`;

exports[`createFormApi submitEvent: submitted 3`] = `false`;

exports[`createFormApi validate 1`] = `
Object {
  "firstName": "",
  "lastName": "",
}
`;

exports[`createFormApi validate 2`] = `
Object {
  "firstName": "foobar",
  "lastName": "",
}
`;

exports[`createFormApi validate: submitted 1`] = `false`;

exports[`createFormApi validate: submitted 2`] = `true`;

exports[`createFormApi validate: submitted 3`] = `false`;

exports[`createFormApi validate: submitted 4`] = `true`;

exports[`createFormApi validate: submitted 5`] = `false`;
